cherrypy/test/test_client_cert.py

```diff
@@ -0,0 +1,140 @@
+"""Tests for client certificate validation."""
+
+import itertools
+import json
+import pytest
+import ssl
+from os.path import abspath, basename, dirname, join
+from six.moves import urllib
+
+import cherrypy
+from cherrypy.test import helper
+
+
+SSL_DIR = join(abspath(dirname(__file__)), 'ssl')
+CA_CERT = join(SSL_DIR, 'ca.cert')
+CLIENT_KEY = join(SSL_DIR, 'client.key')
+CLIENT_CERT = join(SSL_DIR, 'client.cert')
+CLIENT_IP_CERT = join(SSL_DIR, 'client_ip.cert')
+CLIENT_WILD_CERT = join(SSL_DIR, 'client_wildcard.cert')
+CLIENT_WRONG_CA = join(SSL_DIR, 'client_wrong_ca.cert')
+CLIENT_WRONG_HOST = join(SSL_DIR, 'client_wrong_host.cert')
+SERVER_KEY = join(SSL_DIR, 'server.key')
+SERVER_CERT = join(SSL_DIR, 'server.cert')
+
+
+def subdict(d, key_predicate):
+    return {k: v for k, v in d.items() if key_predicate(k)}
+
+
+class ClientCertTests(object):
+    """Client certificate validation tests."""
+    scheme = 'https'
+    script_name = '/wsgi_env'
+
+    server_should_reject = False
+    server_host = 'localhost'
+    client_host = 'localhost'
+    ca_chain = CA_CERT
+    client_cert = CLIENT_CERT
+    verify_mode = None  # None, 'none', 'optional', 'required'
+
+    @classmethod
+    def setup_server(cls):
```

Comment:
How do you run this method exactly?

Reply:
It's run by CPWebCase.setup_class; this class, ClientCertTests, is multi-inherited with CPWebCase into the base test classes at line 137.